### PR TITLE
181162867 author sensor and pre-recoded data

### PR DESCRIPTION
**Pull request by knowuh**
# PR for Initial work (WIP):

- Add unit election to authoring
- Add range & unit preview to authoring
- Add CSV data field to authoring
- Pass down the prerorded data to the runtime
- Reorganize form
- Update form styles

[#181162867]
https://www.pivotaltracker.com/story/show/181162867

### CODE REVIEW

**Comment by knowuh** [comment]
@dougmartin -- We should probably go over this together

**Comment by knowuh** [comment]
@dougmartin I added `src/utils/merge-timeseries-data.ts` and `src/utils/merge-timeseries-data.test.ts` for the data merging issue.  I might push a few more small changes on this branch, but essentially I think its ready for review.

**Comment by knowuh** [comment]
@dougmartin I think this PR is ready, we are interpolating values now, and can graph sensors with prerecorded author data. I think we should get this merged before it gets too large.

File: src/components/sensor-graph.tsx [hunk 1]
```diff
@@ -69,25 +71,42 @@ export default class SensorGraph extends React.Component<SensorGraphProps, Senso
...
-              { yMin, yMax } = this.state,
-              plotYMin = yMin != null ? yMin : (sensorRecording?.min != null ? sensorRecording.min : 0),
-              plotYMax = yMax != null ? yMax : (sensorRecording?.max != null ? sensorRecording.max : 10);
+        const { sensorRecording, preRecording } = this.props;
+        const { yMin, yMax } = this.state;
+        const source = sensorRecording || preRecording;
+        const plotYMin = yMin != null ? yMin : (source?.min != null ? source.min : 0);
+        const plotYMax = yMax != null ? yMax : (source?.max != null ? source.max : 10);
+
+ let data:number[][] = [];
```
Comment (dougmartin) [inline src/components/sensor-graph.tsx:89]: Missing indention here.

File: src/components/sensor-graph.tsx [hunk 2]
```diff
@@ -69,25 +71,42 @@ export default class SensorGraph extends React.Component<SensorGraphProps, Senso
...
+        const hasSensorData = sensorRecording && sensorRecording.data.length > 0;
+        // Dygraph requires each row of data to have the same number of columns.
+        // If we have both the sensor and pre-recording data, plot both.
+        if (preRecording && hasSensorData) {
+            data = mergeTimeSeriesData(preRecording.data, sensorRecording.data);
+            const maxSensorTime = sensorRecording.data[sensorRecording.data.length - 1][0];
```
Comment (dougmartin) [inline src/components/sensor-graph.tsx:95]: So we clip the pre-recording data at the last sensor data time?

File: src/interactive/authoring.tsx [hunk 3]
```diff
@@ -20,31 +23,144 @@ export const AuthoringComponent: React.FC<Props> = ({initMessage}) => {
...
+      let data = [] as number[][];
+      let rows = [];
+      try {
+        rows = value.split('\n');
+        for (let row of rows) {
+          data.push(row.split(/,/g).map(d => parseFloat(d)));
+        }
+      }
+      catch (e) {
+        console.error('cant parse data', e);
```
Comment (dougmartin) [inline src/interactive/authoring.tsx:49]: I'm wondering if we should show some UI element to let the authors know the data can't be parsed?

File: src/interactive/authoring.tsx [hunk 4]
```diff
@@ -20,31 +23,144 @@ export const AuthoringComponent: React.FC<Props> = ({initMessage}) => {
...
+      }
+
+      const unit = authoredState.sensorUnit as string;
+      const definition = SensorDefinitions[unit];
+      const precision = 2;
+      const columnID = "110";   // TODO: ??
+      const sensorPosition = 0; // TODO: ??
```
Comment (dougmartin) [inline src/interactive/authoring.tsx:56]: Are these TODOs still valid?

File: src/components/sensor-graph.tsx [hunk 5]
```diff
@@ -69,25 +71,42 @@ export default class SensorGraph extends React.Component<SensorGraphProps, Senso
...
+        const { yMin, yMax } = this.state;
+        const source = sensorRecording || preRecording;
+        const plotYMin = yMin != null ? yMin : (source?.min != null ? source.min : 0);
+        const plotYMax = yMax != null ? yMax : (source?.max != null ? source.max : 10);
+
+ let data:number[][] = [];
+        const hasSensorData = sensorRecording && sensorRecording.data.length > 0;
+        // Dygraph requires each row of data to have the same number of columns.
+        // If we have both the sensor and pre-recording data, plot both.
+        if (preRecording && hasSensorData) {
```
Comment (dougmartin) [inline src/components/sensor-graph.tsx:93]: The `mergeTimeSeriesData` doesn't handle empty data arrays well and this conditional doesn't check the length of the `preRecording` array.  If it has a 0 length `mergeTimeSeriesData` is going to fail.

File: src/utils/merge-timeseries-data.ts [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+  if (nextTime === time) {
+    return nextValue;
+  }
+  const t = (time - lastTime) / (nextTime - lastTime);
+  return lerp(lastValue, nextValue, t);
+}
+
+// Given two input arrays of time series data, merge them into a single array
+// where each element is a triplet of  numbers representing a time value of a and b.
+export const mergeTimeSeriesData = (dataA: timeSeriesData, dataB:timeSeriesData) => {
```
Comment (dougmartin) [inline src/utils/merge-timeseries-data.ts:33]: I think to be defensive you should check the length of both `dataA` and `dataB` before you start indexing into them

**Review by dougmartin — CHANGES_REQUESTED** [review]
Overall looks good - just a few comments preventing approval.

File: src/components/sensor-graph.tsx (hunk 5 above)
Comment (knowuh) [reply src/components/sensor-graph.tsx:93]: Fixed in [a58a392](https://github.com/concord-consortium/sensor-interactive/pull/80/commits/a58a3927a52fc291015f8ef63f01613772e190a6)

File: src/utils/merge-timeseries-data.ts (hunk 6 above)
Comment (knowuh) [reply src/utils/merge-timeseries-data.ts:33]: Fixed in [a58a392](https://github.com/concord-consortium/sensor-interactive/pull/80/commits/a58a3927a52fc291015f8ef63f01613772e190a6)

**Comment by knowuh** [comment]
@dougmartin -- I think I addressed all your requests, see [a58a392](https://github.com/concord-consortium/sensor-interactive/pull/80/commits/a58a3927a52fc291015f8ef63f01613772e190a6) & [50caca6](https://github.com/concord-consortium/sensor-interactive/pull/80/commits/50caca6809d11f0bf86d285e5fb35b18592d701f)

**Review by dougmartin — APPROVED** [review]
Looks good :-1:

**Pull request merged**